Sidebar extensions should temporarily receive the activeTab permission when the user interacts with the sidebar.

AriYoung00 opened the pull request:
#### 01d2ecafac25cb1c2defa2f65cec3a4ce7dac42d
<pre>
Sidebar extensions should temporarily receive the activeTab permission when the user interacts with the sidebar.
<a href="https://webkit.org/b/279281">https://webkit.org/b/279281</a>
<a href="https://rdar.apple.com/135375961">rdar://135375961</a>

Reviewed by Timothy Hatcher.

This patch introduces changes to WebExtensionSidebar that allow it to
detect when a user opens the sidebar or interacts with the sidebar&apos;s
WKWebView, so that we can grant activeTab to the sidebar&apos;s extension
upon such interactions. It also cleans up some existing code in the same
class that was made unnecessary by this change. This patch also
includes a change that will cause WebKit to post notifications to the
browser when the activeTab permission is granted or remove, so that the
browser can immediately update the extension&apos;s UI/actionIcon to reflect
the presence or absence of this permission. Finally, this patch also
fixes some sidebar tests and fixes WebExtensionSidebar::icon to reflect
recent changes to WebExtension::bestImageInIconsDictionary.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionSidebar.h: Amend doc
  comment to more accurately reflect the purpose of willOpenSidebar.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
(WebKit::WebExtensionContext::openSidebarForTab): Remove userTriggered
parameter, as this method should only be used for programmatic API
opening/closing.
(WebKit::WebExtensionContext::closeSidebarForTab): Remove userTriggered
parameter.
(WebKit::WebExtensionContext::sidebarOpen): Remove userTriggered
parameter from call to openSidebarForTab.
(WebKit::WebExtensionContext::sidebarClose): Remove userTriggered
parameter from call to closeSidebarForTab.
(WebKit::WebExtensionContext::sidebarToggle): Remove userTriggered
parameter from calls to openSidebarForTab and closeSidebarForTab.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::userGesturePerformed): Call
permissionsDidChange when activeTab is granted to post notification to
browser.
(WebKit::WebExtensionContext::clearUserGesture): Call
permissionsDidChange when activeTab is revoked to post notification to
browser.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
(-[_WKWebExtensionSidebarViewController viewWillAppear]): Call
sidebar-&gt;sidebarWillAppear rather than sidebar-&gt;willOpenSidebar.
(-[_WKWebExtensionSidebarViewController viewWillDisappear]): Call
sidebar-&gt;sidebarWillDisappear rather than sidebar-&gt;willCloseSidebar.
(-[_WKWebExtensionSidebarWebView initWithFrame:configuration:webExtensionSidebar:]): Added.
(-[_WKWebExtensionSidebarWebView mouseDown:]): Added -- calls
sidebar-&gt;didReceiveUserInteraction when web view receives left mouse
click events.
(-[_WKWebExtensionSidebarWebView rightMouseDown:]): Added -- calls
sidebar-&gt;didReceiveUserInteraction when web view receives right mouse
click events.
(-[_WKWebExtensionSidebarWebView otherMouseDown:]): Added -- calls
sidebar-&gt;didReceiveUserInteraction when web view receives middle mouse
click events.
(WebKit::WebExtensionSidebar::icon): Update usage of
WebExtension::bestImageInIconsDictionary to reflect new signature.
(WebKit::WebExtensionSidebar::willOpenSidebar): Remove erroneous assert
and log statement, add call to didReceiveUserInteraction.
(WebKit::WebExtensionSidebar::sidebarWillAppear): Added.
(WebKit::WebExtensionSidebar::sidebarWillDisappear): Added.
(WebKit::WebExtensionSidebar::didReceiveUserInteraction): Helper method
to call WebExtensionContext::userGesturePerformed if all necessary
information is present.
(WebKit::WebExtensionSidebar::webView): Instantiate a
_WKWebExtensionSidebarWebView rather than a WKWebView.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h: Remove
  userTriggered parameter from declarations of openSidebarForTab and
closeSidebarForTab.
* Source/WebKit/UIProcess/Extensions/WebExtensionSidebar.h: Add
  declarations of sidebarWillAppear, sidebarWillDisappear, and
didReceiveUserInteraction.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm:
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidePanelAPIGlobalEnablePersists)):
Invert asserts to reflect changed default enablement value.
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidePanelAPITabEnablePersists)):
Invert asserts to reflect changed default enablement value.
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidePanelModifyGlobalEnable)):
Invert asserts to reflect changed default enablement value.

Canonical link: <a href="https://commits.webkit.org/283300@main">https://commits.webkit.org/283300@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fddb000241324e1999b05d21b000e9220ad443ad

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65828 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45201 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18447 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69857 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16437 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67946 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53000 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16719 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52848 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11414 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68895 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41722 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56979 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33480 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38397 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14358 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15313 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60243 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14702 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71560 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9783 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14124 "Found 1 new test failure: http/wpt/mediastream/mediastreamtrackprocessor-videoframe-timestamp.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/60152 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9815 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57046 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60436 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8078 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1724 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9974 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41009 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42085 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43268 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41829 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -403,6 +451,20 @@ - (void)viewWillDisappear
+    if (!(isOpen() && currentTab && currentContext))
+        return;
+
+    extensionContext().value()->userGesturePerformed(currentTab.value());
```
Use `currentContext` here?

AriYoung00 replied:
Whoops, big brain fart there 😅

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -403,6 +451,20 @@ - (void)viewWillDisappear
+    auto currentContext = extensionContext();
+
+    RELEASE_LOG_ERROR_IF(!isOpen(), Extensions, "didReceiveUserInteraction was called on a sidebar object which is not open");
+    RELEASE_LOG_ERROR_IF(!currentTab, Extensions, "didReceiveUserInteraction was called on a sidebar which has no associated tab");
```
Should these just be ASSERTs?

AriYoung00 replied:
They definitely could be, I don't really have a strong preference either way. It seems like this logging could possibly be useful for debugging sidebar issues that could theoretically occur in the wild, but it was more of a "why not log something" situation when I wrote it -- switching them over right now.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -3069,6 +3069,8 @@ static WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
     RefPtr pattern = WebExtensionMatchPattern::getOrCreate(currentURL);
     tab.setTemporaryPermissionMatchPattern(pattern.copyRef());
+    if (pattern)
+        permissionsDidChange(WKWebExtensionContextPermissionMatchPatternsWereGrantedNotification, MatchPatternSet { *pattern });
```
Can you file a bug about passing `tab` to `permissionsDidChange()` so it can be included in the notification and add as a FIXME comment here? I do worry this could be confusing without including it, making it seem like broader permissions were granted.

AriYoung00 committed:
Sidebar extensions should temporarily receive the activeTab permission when the user interacts with the sidebar.
https://webkit.org/b/279281
rdar://135375961

Reviewed by Timothy Hatcher.

This patch introduces changes to WebExtensionSidebar that allow it to
detect when a user opens the sidebar or interacts with the sidebar's
WKWebView, so that we can grant activeTab to the sidebar's extension
upon such interactions. It also cleans up some existing code in the same
class that was made unnecessary by this change. This patch also
includes a change that will cause WebKit to post notifications to the
browser when the activeTab permission is granted or remove, so that the
browser can immediately update the extension's UI/actionIcon to reflect
the presence or absence of this permission. Finally, this patch also
fixes some sidebar tests and fixes WebExtensionSidebar::icon to reflect
recent changes to WebExtension::bestImageInIconsDictionary.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionSidebar.h: Amend doc
  comment to more accurately reflect the purpose of willOpenSidebar.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
(WebKit::WebExtensionContext::openSidebarForTab): Remove userTriggered
parameter, as this method should only be used for programmatic API
opening/closing.
(WebKit::WebExtensionContext::closeSidebarForTab): Remove userTriggered
parameter.
(WebKit::WebExtensionContext::sidebarOpen): Remove userTriggered
parameter from call to openSidebarForTab.
(WebKit::WebExtensionContext::sidebarClose): Remove userTriggered
parameter from call to closeSidebarForTab.
(WebKit::WebExtensionContext::sidebarToggle): Remove userTriggered
parameter from calls to openSidebarForTab and closeSidebarForTab.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::userGesturePerformed): Call
permissionsDidChange when activeTab is granted to post notification to
browser.
(WebKit::WebExtensionContext::clearUserGesture): Call
permissionsDidChange when activeTab is revoked to post notification to
browser.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
(-[_WKWebExtensionSidebarViewController viewWillAppear]): Call
sidebar->sidebarWillAppear rather than sidebar->willOpenSidebar.
(-[_WKWebExtensionSidebarViewController viewWillDisappear]): Call
sidebar->sidebarWillDisappear rather than sidebar->willCloseSidebar.
(-[_WKWebExtensionSidebarWebView initWithFrame:configuration:webExtensionSidebar:]): Added.
(-[_WKWebExtensionSidebarWebView mouseDown:]): Added -- calls
sidebar->didReceiveUserInteraction when web view receives left mouse
click events.
(-[_WKWebExtensionSidebarWebView rightMouseDown:]): Added -- calls
sidebar->didReceiveUserInteraction when web view receives right mouse
click events.
(-[_WKWebExtensionSidebarWebView otherMouseDown:]): Added -- calls
sidebar->didReceiveUserInteraction when web view receives middle mouse
click events.
(WebKit::WebExtensionSidebar::icon): Update usage of
WebExtension::bestImageInIconsDictionary to reflect new signature.
(WebKit::WebExtensionSidebar::willOpenSidebar): Remove erroneous assert
and log statement, add call to didReceiveUserInteraction.
(WebKit::WebExtensionSidebar::sidebarWillAppear): Added.
(WebKit::WebExtensionSidebar::sidebarWillDisappear): Added.
(WebKit::WebExtensionSidebar::didReceiveUserInteraction): Helper method
to call WebExtensionContext::userGesturePerformed if all necessary
information is present.
(WebKit::WebExtensionSidebar::webView): Instantiate a
_WKWebExtensionSidebarWebView rather than a WKWebView.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h: Remove
  userTriggered parameter from declarations of openSidebarForTab and
closeSidebarForTab.
* Source/WebKit/UIProcess/Extensions/WebExtensionSidebar.h: Add
  declarations of sidebarWillAppear, sidebarWillDisappear, and
didReceiveUserInteraction.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm:
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidePanelAPIGlobalEnablePersists)):
Invert asserts to reflect changed default enablement value.
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidePanelAPITabEnablePersists)):
Invert asserts to reflect changed default enablement value.
(TestWebKitAPI::TEST_F(WKWebExtensionAPISidebar, SidePanelModifyGlobalEnable)):
Invert asserts to reflect changed default enablement value.

Canonical link: https://commits.webkit.org/283300@main
